spec/lib/bolt_spec/plans/mock_executor.rb

```diff
@@ -0,0 +1,196 @@
+          return false
+        end
+
+        if @invocation[:arguments] && arguments != @invocation[:arguments]
+          return false
+        end
+
+        if @invocation[:options] && options != @invocation[:options]
+          return false
+        end
+
+        true
+      end
+
+      def call(targets, _task, _arguments, _options)
+        @calls += 1
+        Bolt::ResultSet.new(targets.map do |target|
+          val = @data[target.name] || @data[:default]
+          Bolt::Result.new(target, value: val)
+        end)
+      end
+
+      def assert_called(taskname)
+        satisfied = if @expect
+                      (@expected_calls.nil? && @calls > 0) || @calls == @expected_calls
+                    else
+                      @expected_calls.nil? || @calls <= @expected_calls
+                    end
+        unless satisfied
+          unless (times = @expected_calls)
+            times = @expect ? "at least one" : "any number of"
+          end
+          message = "Expected #{taskname} to be called #{times} times"
+          message += " with targets #{@invocation[:targets]}" if @invocation[:targets]
+          message += " with parameters #{@invocations[:parameters]}" if @invocation[:parameters]
+          raise message
+        end
+      end
+
+      # This changes the stub from an allow to an expect which will validate
+      # that it has been called.
+      def expect_call
+        @expect = true
+        self
+      end
+
+      # Below here are the intended 'public' functions
+
+      # Restricts the stub to only match invocations with
+      # the correct targets
+      def with_targets(targets)
+        targets = [targets] unless targets.is_a? Array
+        @invocation[:targets] = targets.map do |target|
+          if target.is_a? String
+            target
+          else
+            target.name
+          end
+        end
+        self
+      end
+
+      # Restricts the stub to only match invocations with certain parameters
+      # All parameters must match exactly and since arguments and options are
+      # treated differently at the executor this won't work with some '_*' options
+      # TODO: Fix handling of '_*' options probably by breaking them into other helpers
+      def with_params(params)
+        @invocation[:parameters] = params
+        @invocation[:arguments] = params.reject { |k, _v| k.start_with?('_') }
+        @invocation[:options] = params.select { |k, _v| k.start_with?('_') }
+        self
+      end
+
+      # limit the maximum number of times an allow stub may be called or
+      # specify how many times an expect stub must be called.
+      def be_called_times(times)
+        @expected_calls = times
+        self
+      end
+
+      # error if the stub is called at all.
+      def not_be_called
+        @expected_calls = 0
+        self
+      end
+
+      # Set different result values for each target
+      # TODO: how should this interact with multiple calls and always_return
```

Comment:
I can see how this is useful, but it's mixing setting expectations - the targets - with specifying the action to take - what value to return.
IMO we should keep it separate (i.e. no `return_for_targets` method), that should allow us to have a list of stubs each with its own set of expectations/constraints  and a _single_ return value.
Then, there is no ambiguity, the return value from the first matching stub is used/returned. 

To provide the flexibility this offers we should instead look into implementing something like [rspec-mock's block implementation](https://relishapp.com/rspec/rspec-mocks/v/3-7/docs/configuring-responses/block-implementation).

Reply:
I agree. I think we can make a special method that does both if we end up doing this a lot.

At this point I think stubbing tasks is different from stubbing methoss in ruby
1. Task is more likely be called multiple times in a plan in widely different contexts. I think matching on description could help  this
2. I'm not sure rspec-mocks block syntax works for us because we have multiple matches.